ConformanceAKS failed on:
```
command "curl -w %{local_ip}:%{local_port} -> %{remote_ip}:%{remote_port} = %{response_code} --silent --fail --show-error --connect-timeout 5 --output /dev/null [https://1.1.1.1:443](https://1.1.1.1/)" succeeded while it should have failed:
ℹ️  curl output:
 No flows recorded during action cloudflare-1111-0
```

And runtime tests are some panic that doesn't seem related:
```
	 [21:01:10]     testcase.go:144: Failed to start cilium agent: could not build arguments for function "github.com/cilium/cilium/daemon/cmd".glob..func2 (/home/vagrant/go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:1627): failed to build promise.Promise[*github.com/cilium/cilium/daemon/cmd.Daemon]: could not build arguments for function "github.com/cilium/cilium/daemon/cmd".newDaemonPromise (/home/vagrant/go/src/github.com/cilium/cilium/daemon/cmd/daemon_main.go:1642): failed to build *agent.Controller: missing dependencies for function "github.com/cilium/cilium/pkg/bgpv1/agent".NewController (/home/vagrant/go/src/github.com/cilium/cilium/pkg/bgpv1/agent/controller.go:188): missing type: *option.DaemonConfig
	 [21:01:10] --- FAIL: TestControlPlane (0.00s)
	 [21:01:10]     --- FAIL: TestControlPlane/CNPStatusNodesGC (0.00s)
	 [21:01:10]     --- FAIL: TestControlPlane/IdentityGC (0.00s)
	 [21:01:10] panic: runtime error: invalid memory address or nil pointer dereference
	 [21:01:10] 	panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	 [21:01:10] 	panic: runtime error: invalid memory address or nil pointer dereference
	 [21:01:10] [signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x2a9217d]
	 [21:01:10] 
	 [21:01:10] goroutine 157 [running]:
	 [21:01:10] testing.tRunner.func1.2({0x2ea0160, 0x5743740})
	 [21:01:10] 	/usr/local/go/src/testing/testing.go:1396 +0x24e
	 [21:01:10] testing.tRunner.func1()
	 [21:01:10] 	/usr/local/go/src/testing/testing.go:1399 +0x39f
	 [21:01:10] panic({0x2ea0160, 0x5743740})
	 [21:01:10] 	/usr/local/go/src/runtime/panic.go:884 +0x212
	 [21:01:10] github.com/cilium/cilium/test/controlplane/suite.(*agentHandle).tearDown(0x0)
	 [21:01:10] 	/home/vagrant/go/src/github.com/cilium/cilium/test/controlplane/suite/agent.go:33 +0x1d
	 [21:01:10] github.com/cilium/cilium/test/controlplane/suite.(*ControlPlaneTest).StopAgent(0xc000cc3ea0)
	 [21:01:10] 	/home/vagrant/go/src/github.com/cilium/cilium/test/controlplane/suite/testcase.go:152 +0x25
	 [21:01:10] panic({0x2ea0160, 0x5743740})
	 [21:01:10] 	/usr/local/go/src/runtime/panic.go:884 +0x212
	 [21:01:10] github.com/cilium/cilium/test/controlplane/suite.(*operatorHandle).tearDown(0x0)
	 [21:01:10] 	/home/vagrant/go/src/github.com/cilium/cilium/test/controlplane/suite/operator.go:19 +0x19
	 [21:01:10] github.com/cilium/cilium/test/controlplane/suite.(*ControlPlaneTest).StopOperator(0xc000cc3ea0)
	 [21:01:10] 	/home/vagrant/go/src/github.com/cilium/cilium/test/controlplane/suite/testcase.go:190 +0x25
	 [21:01:10] runtime.Goexit()
	 [21:01:10] 	/usr/local/go/src/runtime/panic.go:522 +0x185
	 [21:01:10] testing.(*common).FailNow(0xc000d02000)
	 [21:01:10] 	/usr/local/go/src/testing/testing.go:871 +0x4c
	 [21:01:10] testing.(*common).Fatalf(0xc000d02000, {0x34cf06a?, 0x347902f?}, {0xc000d5de70?, 0xc000cad980?, 0x30?})
	 [21:01:10] 	/usr/local/go/src/testing/testing.go:955 +0x65
	 [21:01:10] github.com/cilium/cilium/test/controlplane/suite.(*ControlPlaneTest).StartAgent(0xc000cc3ea0)
	 [21:01:10] 	/home/vagrant/go/src/github.com/cilium/cilium/test/controlplane/suite/testcase.go:144 +0x1bb
	 [21:01:10] github.com/cilium/cilium/test/controlplane/identities.init.0.func1(0x0?)
	 [21:01:10] 	/home/vagrant/go/src/github.com/cilium/cilium/test/controlplane/identities/identity_gc.go:147 +0xea
	 [21:01:10] testing.tRunner(0xc000d02000, 0x3649998)
	 [21:01:10] 	/usr/local/go/src/testing/testing.go:1446 +0x10b
	 [21:01:10] created by testing.(*T).Run
	 [21:01:10] 	/usr/local/go/src/testing/testing.go:1493 +0x35f
	 [21:01:10] FAIL	github.com/cilium/cilium/test/controlplane	0.070s
	 [21:01:15] OK: 1 passed
```

--- a/pkg/hubble/metrics/api/api.go
+++ b/pkg/hubble/metrics/api/api.go
@@ -5,9 +5,11 @@ package api
 
 import (
 	"context"
+	"fmt"
 	"strings"
 
 	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sirupsen/logrus"
 	"go.uber.org/multierr"
 
 	pb "github.com/cilium/cilium/api/v1/flow"
@@ -39,8 +41,11 @@ func ParseMetricList(enabledMetrics []string) (m Map) {
 	return
 }
 
-// Handlers is a slice of metric handler
-type Handlers []Handler
+// Handlers contains all the metrics handlers.
+type Handlers struct {
+	handlers       []Handler
+	flowProcessors []FlowProcessor
+}
 
 // Plugin is a metric plugin. A metric plugin is associated a name and is
 // responsible to spawn metric handlers of a certain type.
@@ -61,29 +66,50 @@ type PluginConflicts interface {
 	ConflictingPlugins() []string
 }
 
-// Handler is a metric handler. It is called upon receival of raw event data
-// and is responsible to perform metrics accounting according to the scope of
-// the metrics plugin.
+// Handler is a basic metric handler.
 type Handler interface {
 	// Init must initialize the metric handler by validating and parsing
 	// the options and then registering all required metrics with the
 	// specifies Prometheus registry
 	Init(registry *prometheus.Registry, options Options) error
 
+	// Status returns the configuration status of the metric handler
+	Status() string
+}
+
+// FlowProcessor is a metric handler which requires flows to perform metrics
+// accounting.
+// It is called upon receival of raw event data and is responsible
+// to perform metrics accounting according to the scope of the metrics plugin.
+type FlowProcessor interface {
 	// ProcessFlow must processes a flow event and perform metrics
 	// accounting
 	ProcessFlow(ctx context.Context, flow *pb.Flow) error
+}
 
-	// Status returns the configuration status of the metric handler
-	Status() string
+func NewHandlers(log logrus.FieldLogger, registry *prometheus.Registry, in []NamedHandler) (*Handlers, error) {
+	var handlers Handlers
+	for _, item := range in {
+		handlers.handlers = append(handlers.handlers, item.Handler)
+		if fp, ok := item.Handler.(FlowProcessor); ok {
+			handlers.flowProcessors = append(handlers.flowProcessors, fp)
+		}
+
+		if err := item.Handler.Init(registry, item.Options); err != nil {
+			return nil, fmt.Errorf("unable to initialize metric '%s': %s", item.Name, err)
+		}
+
+		log.WithFields(logrus.Fields{"name": item.Name, "status": item.Handler.Status()}).Info("Configured metrics plugin")
+	}
+	return &handlers, nil
 }
 
 // ProcessFlow processes a flow by calling ProcessFlow it on to all enabled
 // metric handlers
 func (h Handlers) ProcessFlow(ctx context.Context, flow *pb.Flow) error {
 	var processingErr error
-	for _, mh := range h {
-		err := mh.ProcessFlow(ctx, flow)
+	for _, fp := range h.flowProcessors {
+		err := fp.ProcessFlow(ctx, flow)
 		// Continue running the remaining metrics handlers, since one failing
 		// shouldn't impact the other metrics handlers.
 		processingErr = multierr.Append(processingErr, err)

--- a/pkg/hubble/metrics/api/registry.go
+++ b/pkg/hubble/metrics/api/registry.go
@@ -38,14 +38,20 @@ func (r *Registry) Register(name string, p Plugin) {
 	r.mutex.Unlock()
 }
 
+type NamedHandler struct {
+	Name    string
+	Options Options
+	Handler Handler
+}
+
 // ConfigureHandlers enables a set of metric handlers and initializes them.
 // Only metrics handlers which have been previously registered via the
 // Register() function can be configured.
-func (r *Registry) ConfigureHandlers(registry *prometheus.Registry, enabled Map) (Handlers, error) {
+func (r *Registry) ConfigureHandlers(registry *prometheus.Registry, enabled Map) (*Handlers, error) {
 	r.mutex.Lock()
 	defer r.mutex.Unlock()
 
-	initialized := make(Handlers, 0, len(enabled))
+	var enabledHandlers []NamedHandler
 	for name, opts := range enabled {
 		plugin, ok := r.handlers[name]
 		if !ok {
@@ -60,14 +66,13 @@ func (r *Registry) ConfigureHandlers(registry *prometheus.Registry, enabled Map)
 			}
 		}
 
-		handler := plugin.NewHandler()
-		if err := handler.Init(registry, opts); err != nil {
-			return nil, fmt.Errorf("unable to initialize metric '%s': %s", name, err)
+		h := NamedHandler{
+			Name:    name,
+			Options: opts,
+			Handler: plugin.NewHandler(),
 		}
-		r.log.WithFields(logrus.Fields{"name": name, "status": handler.Status()}).Info("Configured metrics plugin")
-
-		initialized = append(initialized, handler)
+		enabledHandlers = append(enabledHandlers, h)
 	}
 
-	return initialized, nil
+	return NewHandlers(r.log, registry, enabledHandlers)
 }

--- a/pkg/hubble/metrics/api/registry_test.go
+++ b/pkg/hubble/metrics/api/registry_test.go
@@ -53,13 +53,13 @@ func TestRegister(t *testing.T) {
 
 	handlers, err := r.ConfigureHandlers(nil, Map{})
 	assert.EqualValues(t, err, nil)
-	assert.EqualValues(t, len(handlers), 0)
+	assert.EqualValues(t, len(handlers.handlers), 0)
 
 	handlers, err = r.ConfigureHandlers(nil, Map{"test": Options{}})
 	assert.EqualValues(t, err, nil)
-	assert.EqualValues(t, len(handlers), 1)
-	assert.EqualValues(t, handlers[0].(*testHandler).InitCalled, 1)
+	assert.EqualValues(t, len(handlers.handlers), 1)
+	assert.EqualValues(t, handlers.handlers[0].(*testHandler).InitCalled, 1)
 
 	handlers.ProcessFlow(context.TODO(), nil)
-	assert.EqualValues(t, handlers[0].(*testHandler).ProcessCalled, 1)
+	assert.EqualValues(t, handlers.handlers[0].(*testHandler).ProcessCalled, 1)
 }

--- a/pkg/hubble/metrics/http/handler_test.go
+++ b/pkg/hubble/metrics/http/handler_test.go
@@ -14,6 +14,7 @@ import (
 	"github.com/stretchr/testify/require"
 
 	pb "github.com/cilium/cilium/api/v1/flow"
+	"github.com/cilium/cilium/pkg/hubble/metrics/api"
 )
 
 func Test_httpHandler_Status(t *testing.T) {
@@ -31,15 +32,17 @@ func Test_httpHandler_ProcessFlow(t *testing.T) {
 	handler := plugin.NewHandler()
 	require.Error(t, handler.Init(prometheus.NewRegistry(), map[string]string{"destinationContext": "invalid"}))
 	require.NoError(t, handler.Init(prometheus.NewRegistry(), nil))
+	fp, ok := handler.(api.FlowProcessor)
+	require.True(t, ok)
 	// shouldn't count
-	handler.ProcessFlow(ctx, &pb.Flow{})
+	fp.ProcessFlow(ctx, &pb.Flow{})
 	// shouldn't count
-	handler.ProcessFlow(ctx, &pb.Flow{L7: &pb.Layer7{
+	fp.ProcessFlow(ctx, &pb.Flow{L7: &pb.Layer7{
 		Type:   pb.L7FlowType_RESPONSE,
 		Record: &pb.Layer7_Dns{},
 	}})
 	// should count for request
-	handler.ProcessFlow(ctx, &pb.Flow{
+	fp.ProcessFlow(ctx, &pb.Flow{
 		TrafficDirection: pb.TrafficDirection_INGRESS,
 		L7: &pb.Layer7{
 			Type: pb.L7FlowType_REQUEST,
@@ -49,7 +52,7 @@ func Test_httpHandler_ProcessFlow(t *testing.T) {
 		},
 	})
 	// should count for response
-	handler.ProcessFlow(ctx, &pb.Flow{
+	fp.ProcessFlow(ctx, &pb.Flow{
 		TrafficDirection: pb.TrafficDirection_INGRESS,
 		L7: &pb.Layer7{
 			Type:      pb.L7FlowType_RESPONSE,
@@ -104,17 +107,21 @@ func Test_httpHandlerV2_ProcessFlow(t *testing.T) {
 		"destinationContext": "pod",
 		"labelsContext":      "source_pod,destination_pod",
 	}))
+
+	fp, ok := handler.(api.FlowProcessor)
+	require.True(t, ok)
+
 	// shouldn't count
-	handler.ProcessFlow(ctx, &pb.Flow{})
+	fp.ProcessFlow(ctx, &pb.Flow{})
 	// shouldn't count
-	handler.ProcessFlow(ctx, &pb.Flow{
+	fp.ProcessFlow(ctx, &pb.Flow{
 		TrafficDirection: pb.TrafficDirection_INGRESS,
 		L7: &pb.Layer7{
 			Type:   pb.L7FlowType_RESPONSE,
 			Record: &pb.Layer7_Dns{},
 		}})
 	// shouldn't count for request, we use responses in v2
-	handler.ProcessFlow(ctx, &pb.Flow{
+	fp.ProcessFlow(ctx, &pb.Flow{
 		TrafficDirection: pb.TrafficDirection_INGRESS,
 		L7: &pb.Layer7{
 			Type: pb.L7FlowType_REQUEST,
@@ -147,7 +154,7 @@ func Test_httpHandlerV2_ProcessFlow(t *testing.T) {
 		},
 	}
 	// should count for request
-	handler.ProcessFlow(ctx, &pb.Flow{
+	fp.ProcessFlow(ctx, &pb.Flow{
 		TrafficDirection: pb.TrafficDirection_INGRESS,
 		// Responses have the source and destination inverted, because it's the
 		// other side of the flow. Our tests are asserting that the HTTPv2 handler

--- a/pkg/hubble/metrics/metrics.go
+++ b/pkg/hubble/metrics/metrics.go
@@ -28,7 +28,7 @@ import (
 )
 
 var (
-	enabledMetrics api.Handlers
+	enabledMetrics *api.Handlers
 	registry       = prometheus.NewPedanticRegistry()
 )
 

